# 100318577 - Data Dictionary for EmpData

Disclaimer: The following code for Data Dictonary generation was supported by Microsoft Copilot, an AI-powered assistant to provide assistance in python code generation and debugging.

In [1]:
# LOAD PACKAGES ----
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.gridspec import GridSpec
from matplotlib.ticker import FuncFormatter

# Suppress warnings
import warnings # reduce cluttered output
warnings.filterwarnings("ignore")

### LOAD DATA

In [2]:
# Read the csv file
data = pd.read_csv('EmpData.csv')

# Summarize each column with unique values
unique_summary = {col: data[col].unique().tolist() for col in data.columns}

# Convert the unique summary to a dataframe
unique_summary_df = pd.DataFrame([
    {'Variable': col, 'Unique_Values': ', '.join(map(str, unique_summary[col]))}
    for col in data.columns
])

# Print the unique results summary in a table format
print(unique_summary_df)
print(data.describe())
print(data.info())

         Variable                                      Unique_Values
0       Timestamp  4/27/2021 11:02:10, 4/27/2021 11:02:22, 4/27/2...
1             Age  25-34, 45-54, 35-44, 18-24, 65 or over, 55-64,...
2        Industry  Education (Higher Education), Computing or Tec...
3           Title  Research and Instruction Librarian, Change & I...
4         Context  nan, High school, FT, Data developer/ETL Devel...
5          Salary  55000, 54600, 34000, 62000, 60000, 50000, 1120...
6         Surplus  0.0, 4000.0, nan, 3000.0, 7000.0, 10000.0, 500...
7        Currency  USD, GBP, CAD, EUR, AUD/NZD, Other, CHF, ZAR, ...
8           Other  nan, INR, Peso Argentino, $76,302.34, My bonus...
9     Income_info  nan, I work for a Charter School, Commission b...
10        Country  United States, United Kingdom, US, USA, Canada...
11      State_USA  Massachusetts, nan, Tennessee, Wisconsin, Sout...
12           City  Boston, Cambridge, Chattanooga, Milwaukee, Gre...
13     Experience  5-7 years, 8 - 

### Make data dictionary in a tidy table format

In [15]:
# Define column descriptions
descriptions = {
    'Timestamp': 'The date and time when the data was recorded',
    'Age': 'Age group of individuals',
    'Industry': 'Industry of employment',
    'Title': 'Job title of individuals',
    'Context': 'Work context',
    'Salary': 'Annual salary in USD',
    'Surplus': 'Surplus amount in USD',
    'Currency': 'Currency of salary',
    'Other': 'Other relevant information',
    'Income_info': 'Additional income information',
    'Country': 'Country of residence',
    'State_USA': 'State in the USA',
    'City': 'City of residence',
    'Experience': 'Years of experience',
    'In_Field': 'Years of experience in the field',
    'Education': 'Highest educational qualification',
    'Gender': 'Gender of individuals',
    'Race': 'Race or ethnicity of individuals',
    'Qualification': 'Qualification level'
}

# Define data types
data_types = {col: str(data[col].dtype) for col in data.columns}

# Combine all information into a data dictionary
data_dictionary_df = pd.DataFrame({
    'Variable': data.columns,
    'Description': [descriptions[col] for col in data.columns],
    'Data Type': [data_types[col] for col in data.columns],
    'Unique_Values': [', '.join(map(str, data[col].unique())) for col in data.columns]
})

# Print the data dictionary
# print(data_dictionary_df.to_string(index=False)) # unable to generate pdf when running

| Variable     | Data Type | Description                                  | Example Values                                                   |
|--------------|-----------|----------------------------------------------|------------------------------------------------------------------|
| Timestamp    | String    | The date and time when the data was recorded | 4/27/2021 11:02:10, 4/27/2021 11:02:22, 4/27/2021 11:02:33       |
| Age          | String    | Age group of individuals                     | 18-24, 25-34, 35-44, 45-54, 55-64, 65 or over                    |
| Industry     | String    | Industry of employment                       | Computing or Technology, Education (Higher Education)            |
| Title        | String    | Job title of individuals                     | Change & Innovation, Research and Instruction Librarian          |
| Context      | String    | Work context                                 | Data developer/ETL Developer, FT, High school, nan               |
| Salary       | Float     | Annual salary in USD                         | 34000, 50000, 54600, 55000, 60000, 62000, 112000                 |
| Surplus      | Float     | Surplus amount in USD                        | 0.0, 3000.0, 4000.0, 500.0, 7000.0, 10000.0, nan                 |
| Currency     | String    | Currency of salary                           | AUD/NZD, CAD, CHF, EUR, GBP, Other, USD, ZAR                     |
| Other        | String    | Other relevant information                   | $76,302.34, INR, My bonus..., Peso Argentino, nan                |
| Income_info  | String    | Additional income information                | Commission based..., I work for a Charter School, nan            |
| Country      | String    | Country of residence                         | Canada, United Kingdom, United States, US, USA                   |
| State_USA    | String    | State in the USA                             | Massachusetts, South..., Tennessee, Wisconsin, nan               |
| City         | String    | City of residence                            | Boston, Cambridge, Chattanooga, Gre..., Milwaukee                |
| Experience   | String    | Years of experience                          | 2-4 years, 5-7 years, 8-10 years, 21-30 years                    |
| In_Field     | String    | Years of experience in the field             | 2-4 years, 5-7 years, 11-20 years, 21-30 years                   |
| Education    | String    | Highest educational qualification            | College degree, Master degree, PhD, Some..., nan                 |
| Gender       | String    | Gender of individuals                        | Man, Non-binary, Other or prefer not to say, Woman, nan          |
| Race         | String    | Race or ethnicity of individuals             | Hispanic, Latino or Spanish origin, White, nan                   |
| Qualification| Float     | Qualification level                          | 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, nan                                |


________________________________________________________________________________________________________________________________________

## Catigorical and Continious variables

In [12]:
CATEGORICAL_COLUMNS = ["Industry", "Title", "Context", "Currency", "Other", "Income_info", "Country", 
                       "State_USA", "City", "Experience", "In_Field", "Education", "Gender", "Race"]
CONTINUOUS_COLUMNS = ["Salary", "Surplus", "Qualification"]

print(CATEGORICAL_COLUMNS)
print(CONTINUOUS_COLUMNS)

['Industry', 'Title', 'Context', 'Currency', 'Other', 'Income_info', 'Country', 'State_USA', 'City', 'Experience', 'In_Field', 'Education', 'Gender', 'Race']
['Salary', 'Surplus', 'Qualification']


________________________________________________________________________________________________________________________________________

## Descriptive statistics

In [14]:
# make new dataframe emp
emp = data

# Descriptive statistics for continuous variables
emp['Qualification'] = emp['Qualification'].round(0) # Round 'Qualification' column to 0 decimal places

# Round 'Salary' and 'Surplus' columns to 2 decimal places
emp['Salary'] = emp['Salary'].round(2)
emp['Surplus'] = emp['Surplus'].round(2)

# Descriptive statistics for continuous variables
StatsNum= emp[CONTINUOUS_COLUMNS].describe()

# Descriptive statistics for continuous variables
for column in CONTINUOUS_COLUMNS:
    StatsNum.loc['median', column] = emp[column].median()
    StatsNum.loc['missing', column] = emp[column].isnull().sum()

# Format the statistics to avoid showing .000000 and display 0 for zero values
StatsNum = StatsNum.applymap(lambda x: '{:.2f}'.format(x).rstrip('0').rstrip('.') if x != 0 else '0')

print(StatsNum)

              Salary    Surplus Qualification
count          26713      19762         26502
mean       374952.85    18353.1          3.72
std      37126077.97  854895.36           1.2
min                0          0             1
25%            54000          0             3
50%            75000       2000             3
75%           109000      10000             5
max       6000070000  120000000             6
median         75000       2000             3
missing            0       6951           211


________________________________________________________________________________________________________________________________________